README.md
# USBGadget

Configure USB Gadget functions on your Nerves-based embedded device using
ConfigFS.

## 🚨 This project is work-in-progress 🚨

The intention is that it will become a dependency of [`nerves_init_gadget`] to
allow for automatic configuration of USB Gadget devices as well as an API for
developers to configure their own custom USB Gadget configurations.

[nerves_init_gadget]: https://github.com/nerves-project/nerves_init_gadget

## Installation

To use it directly in your project (i.e. without `nerves_init_gadget`, add
`usb_gadget` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:usb_gadget, "~> 0.1.0"}
  ]
end
```

Documentation can be found at
[https://hexdocs.pm/usb_gadget](https://hexdocs.pm/usb_gadget).



CHANGELOG.md
# Changelog

## v0.1.0

Initial release.

* New Features
  * Configure USB Gadget devices via ConfigFS
  * Pre-configured RNDIS+ECM+ACM composite device


test/usb_gadget_test.exs
defmodule USBGadgetTest do
  use ExUnit.Case
end


mix.exs
defmodule UsbGadget.MixProject do
  use Mix.Project

  def project do
    [
      app: :usb_gadget,
      version: "0.1.0",
      elixir: "~> 1.6",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      description: description(),
      docs: docs(),
      package: package()
    ]
  end

  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp deps do
    [
      {:ex_doc, "~> 0.18.0", only: :dev}
    ]
  end

  defp description do
    """
    Configure USB Gadget devices in Linux
    """
  end

  defp docs do
    [extras: ["README.md"], main: "readme"]
  end

  defp package do
    [
      maintainers: ["Greg Mefford", "Tim Mecklem", "Connor Rigby"],
      files: ["lib", "LICENSE", "mix.exs", "README.md"],
      licenses: ["Apache 2.0"],
      links: %{"Github" => "https://github.com/nerves-project/usb_gadget"}
    ]
  end
end


lib/usb_gadget/builtin.ex
defmodule USBGadget.Builtin do
  @moduledoc """
  These convenience functions allow you to quickly set up some pre-configured
  USB Gadget devices for common use-cases. The devices are configured but not
  enabled automatically, allowing them to be customized as needed.
  """

  import USBGadget

  @doc """
  Creates a USB Gadget device with RNDIS Ethernet, ECM Ethernet, and ACM
  serial interfaces.

  ## Examples
      iex> USBGadget.Builtin.create_rndis_ecm_acm("g")
      :ok
      iex> USBGadget.enable_device("g")
      :ok
  """
  def create_rndis_ecm_acm(device_name) do
    # * bcdUSB 0x200 means USB 2.0
    # * bDevice{Class,SubClass,Protocol} = {0xEF, 0x02, 0x01} means
    #   Interface Association Descriptor class
    # * idVendor and idProduct are reserved using http://pid.codes
    # * bcdDevice is the "firmware version" in binary-coded decimal
    #   so 0x0123 means 1.2.3
    # * These os_desc options tell Windows to use configuration 1
    # * strings/0x409 means EN-US.
    create_device(device_name, %{
      "bcdUSB" => "0x0200",
      "bDeviceClass" => "0xEF",
      "bDeviceSubClass" => "0x02",
      "bDeviceProtocol" => "0x01",
      "idVendor" => "0x1209",
      "idProduct" => "0x0070",
      "bcdDevice" => "0x0100",
      "os_desc" => %{
        "use" => "1",
        "b_vendor_code" => "0xcd",
        "qw_sign" => "MSFT100"
      },
      "strings" => %{
        "0x409" => %{
          "manufacturer" => "Nerves Project",
          "product" => "Ethernet + Serial Gadget",
          "serialnumber" => ""
        }
      }
    })

    # This sub_compatible_id tells Windows to use the RNDIS driver version that
    # actually works instead of a newer one that doesn't. It's not well-
    # documented, but it's commonly used for Linux-based USB Gadgets.
    create_function(device_name, "rndis.usb0", %{
      "os_desc" => %{
        "interface.rndis" => %{
          "compatible_id" => "RNDIS",
          "sub_compatible_id" => "5162001"
        }
      }
    })

    create_function(device_name, "ecm.usb1", %{})

    create_function(device_name, "acm.GS0")

    # * bmAttributes is a bitmap field:
    #   * bit 7 (MSB) is reserved in USB 2.0 and set to 1
    #   * bit 6 set means the device is self-powered
    #   * bit 5 set means the device can wake the host from suspend
    #   * bits 4..0 are reserved and set to 0
    # * MaxPower is the max current the device will draw (in mA)
    create_config(device_name, "c.1", %{
      "bmAttributes" => "0xC0",
      "MaxPower" => "500",
      "strings" => %{
        "0x409" => %{
          "configuration" => "RNDIS and ECM Ethernet with ACM Serial"
        }
      }
    })

    link_functions(device_name, "c.1", ["rndis.usb0", "ecm.usb1", "acm.GS0"])
    link_os_desc(device_name, "c.1")
  end
end


lib/usb_gadget.ex
defmodule USBGadget do
  @moduledoc """
  Some good documentation about ConfigFS can be found here:
  * <http://events.linuxfoundation.org/sites/events/files/slides/USB%20Gadget%20Configfs%20API_0.pdf>
  * <https://www.kernel.org/doc/Documentation/usb/gadget_configfs.txt>
  * <https://www.kernel.org/doc/Documentation/filesystems/configfs/configfs.txt>
  * <https://docs.microsoft.com/en-us/windows-hardware/drivers/usbcon/usb-interface-association-descriptor>
  * <http://www.usb.org/developers/docs/whitepapers/iadclasscode_r10.pdf>
  """

  require Logger

  @gadget_root Application.get_env(:usb_gadget, :root, "/sys/kernel/config/usb_gadget")

  @doc """
  Define a USB Gadget device

  ## Examples

      iex> USBGadget.create_device("g", %{
        "bcdUSB" => "0x0200",
        "bDeviceClass" => "0xEF",
        "bDeviceSubClass" => "0x02",
        "bDeviceProtocol" => "0x01",
        "idVendor" => "0xDEAD",
        "idProduct" => "0xBEEF",
        "bcdDevice" => "0x0100",
        "os_desc" => %{
          "use" => "1",
          "b_vendor_code" => "0xcd",
          "qw_sign" => "MSFT100"
        },
        "strings" => %{
          "0x409" => %{
            "manufacturer" => "ACME"
            "product" => "USB-O-Tron 9001"
            "serialnumber" => "1"
          }
        }
      })
      :ok
  """
  def create_device(device_name, attrs \\ %{}) do
    set_attribute(@gadget_root, {device_name, attrs})
    :ok
  end

  @doc """
  Define a function the device will support

  ## Examples

      iex> USBGadget.create_function("g", "rndis.usb0", %{
        "dev_addr" => "de:ad:be:ef:00:00",
        "host_addr" => "de:ad:be:ef:00:02",
        "os_desc" => %{
          "interface.rndis" => %{
            "compatible_id" => "RNDIS",
            "sub_compatible_id" => "5162001"
          }
        }
      })
      :ok
  """
  def create_function(device_name, function_name, attrs \\ %{}) do
    [@gadget_root, device_name, "functions"]
    |> Path.join()
    |> set_attribute({function_name, attrs})

    :ok
  end

  @doc """
  Define a config for the device

  ## Examples

      iex> USBGadget.create_config("g", "c.1", %{
        "bmAttributes" => "0xC0",
        "MaxPower" => "1",
        "strings" => %{
          "0x409" => %{
            "configuration" => "Ethernet and Serial"
          }
        }
      })
      :ok
  """
  def create_config(device_name, config_name, attrs \\ %{}) do
    [@gadget_root, device_name, "configs"]
    |> Path.join()
    |> set_attribute({config_name, attrs})

    :ok
  end

  @doc """
  Define the functions available for a given configuration of a device

  ## Examples

      iex> USBGadget.link_functions("g", "c.1", ["rndis.usb0", "ecm.usb1", "acm.GS0"])
      :ok
  """
  def link_functions(device_name, config_name, function_names) do
    function_root = Path.join([@gadget_root, device_name, "functions"])
    config_path = Path.join([@gadget_root, device_name, "configs", config_name])

    Enum.each(function_names, fn function_name ->
      Logger.debug("Linking function #{function_name} > #{config_path}")
      File.ln_s!(Path.join(function_root, function_name), Path.join(config_path, function_name))
    end)

    :ok
  end

  @doc """
  Define the operating system descriptor to a configuration

  ## Examples

      iex> USBGadget.link_os_desc("g", "c.1")
      :ok
  """
  def link_os_desc(device_name, config_name) do
    Logger.debug("Linking os_desc #{device_name} > #{config_name}")
    config_path = Path.join([@gadget_root, device_name, "configs", config_name])
    os_desc_path = Path.join([@gadget_root, device_name, "os_desc", config_name])
    File.ln_s!(config_path, os_desc_path)
    :ok
  end

  @doc """
  Enable the gadget device, making the configured functions available to the operating system

  ## Examples

      iex> USBGadget.enable_device("g")
      :ok
  """
  def enable_device(device_name) do
    [udc_name | _] = File.ls!("/sys/class/udc")
    device_path = udc_path(device_name)
    Logger.debug("Enabling device #{udc_name} > #{device_path}")
    File.write!(device_path, udc_name)
    :ok
  end

  @doc """
  Disable the gadget device, making the configured functions no longer available to the operating system

  ## Examples

      iex> USBGadget.disable_device("g")
      :ok
  """
  def disable_device(device_name) do
    Logger.debug("Disabling device #{device_name}")

    device_name
    |> udc_path()
    |> File.write!("")

    :ok
  end

  # Private Helpers

  defp set_attribute(path, {name, %{} = values}) do
    attr_path = Path.join(path, name)
    File.mkdir_p!(attr_path)
    Enum.each(values, &set_attribute(attr_path, &1))
  end

  defp set_attribute(path, {name, value}) do
    Logger.debug("Setting attribute #{path}/#{name} => #{to_string(value)}")

    :ok =
      path
      |> Path.join(name)
      |> File.write(to_string(value))
  end

  defp udc_path(device_name), do: Path.join([@gadget_root, device_name, "UDC"])
end
